### ISSUE TITLE
Subprocess operations in jupyter kernel fails

### ISSUE THREAD

**Issue by chronitis** [xonsh/xonsh#2044]
With 0.5.1, `xonsh.jupyter_kernel` hangs when trying to carry out any operation with a subprocess:

(The result of expressions is also doubly returned as an execution result and stdout, but that is probably a separate issue).

```
$ XONSH_DEBUG=1 jupyter console --kernel xonsh
In [1]: 1+2+3
Out[1]: 6
6
In [2]: !(/bin/false).returncode
Out[2]: 1
1
In [3]: ls
[IPKernelApp] ERROR | Exception in message handler:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/xonsh/base_shell.py", line 319, in default
    run_compiled_code(code, self.ctx, None, 'single')
  File "/usr/lib/python3/dist-packages/xonsh/codecache.py", line 65, in run_compiled_code
    func(code, glb, loc)
  File "<xonsh-code>", line 1, in <module>
  File "/usr/lib/python3/dist-packages/ipykernel/displayhook.py", line 34, in __call__
    sys.stdout.flush()
  File "/usr/lib/python3/dist-packages/xonsh/base_shell.py", line 170, in flush
    self.std.flush()
  File "/usr/lib/python3.5/tempfile.py", line 848, in flush
    self._file.flush()
ValueError: I/O operation on closed file.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/ipykernel/kernelbase.py", line 228, in dispatch_shell
    handler(stream, idents, msg)
  File "/usr/lib/python3/dist-packages/ipykernel/kernelbase.py", line 390, in execute_request
    user_expressions, allow_stdin)
  File "/usr/lib/python3/dist-packages/xonsh/jupyter_kernel.py", line 55, in do_execute
    shell.default(code)
  File "/usr/lib/python3/dist-packages/xonsh/base_shell.py", line 328, in default
    print_exception()
  File "/usr/lib/python3/dist-packages/xonsh/tools.py", line 630, in print_exception
    sys.stderr.write('xonsh: For full traceback set: '
  File "/usr/lib/python3/dist-packages/xonsh/base_shell.py", line 165, in write
    self.std.write(std_s)
  File "/usr/lib/python3.5/tempfile.py", line 905, in write
    rv = file.write(s)
ValueError: I/O operation on closed file.
```

After which the process hangs and has to be killed.

(edit: updated traceback with non-amalgam version)

**Comment by chronitis** [comment]
This varies for the different subprocess capturing operators:

 * `!(ls)`, `$[ls]` work as expected
 * `ls`, `$(ls)`, `![ls]` generate the above traceback

**Comment by chronitis** [comment]
Possible cause: `xonsh.built_ins.run_subproc()` only calls `command.end()` (`built_ins.py:808`) for `captured='stdout'` or `captured='hiddenobject'`, corresponding to the success/fail categories above.

`xonsh.proc.CommandPipeline.end()` calls `_close_proc()` which closes the file descriptors associated with the process, which presumably ends up with the kernel's stdout temporary file getting closed too early.

I'm afraid I'm not quite confident enough with the internals of xonsh's subprocess mechanisms to know where to fix this.

**Comment by scopatz** [comment]
Thanks for reporting this @chronitis! Sorry for the delay. We'll get to this ASAP, but it might not be until this weekend or next week.

**Comment by scopatz** [comment]
This has been fixed in the v0.5.3 release. Please let us know if you continue to have issues!
